File: src/smscsim/Pdu.java
package smscsim;

import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

class Pdu {

  protected final SmppCmd cmd;
  protected final SmppSts sts;
  protected final int seq;

  Pdu(SmppCmd cmd, SmppSts sts, int seq) {
    this.cmd = cmd;
    this.sts = sts;
    this.seq = seq;
  }

  byte[] toBytes() {
    throw new UnsupportedOperationException("unnecessary operation not implemented in smsc");
  }

  void parseBody(ByteBuffer buf) {
    throw new UnsupportedOperationException("unnecessary operation not implemented in smsc");
  }
}

class HeaderPdu extends Pdu {

  HeaderPdu(SmppCmd cmd, SmppSts sts, int seq) {
    super(cmd, sts, seq);
  }

  HeaderPdu(SmppCmd cmd, int seq) {
    this(cmd, SmppSts.OK, seq);
  }

  @Override
  byte[] toBytes() {
    var buf = ByteBuffer.allocate(16);
    buf.putInt(16);
    buf.putInt(cmd.getCmdId());
    buf.putInt(sts.getStsId());
    buf.putInt(seq);
    return buf.array();
  }

  @Override
  void parseBody(ByteBuffer buf) {
    // noop
  }
}

class BindReqPdu extends Pdu {

  private String systemId = "";

  BindReqPdu(SmppCmd cmd, SmppSts sts, int seq) {
    super(cmd, sts, seq);
  }

  @Override
  void parseBody(ByteBuffer buf) {
    systemId = Utils.parseCStr(buf);
  }

  String systemId() {
    return systemId;
  }

  Pdu createBindResp(SmppSts sts, int seq, String systemId) {
    int respCmdId = cmd.getCmdId() + 0x80000000; // a little hack to calc resp cmd id
    var bindResp = new BindRespPdu(SmppCmd.fromCmdId(respCmdId), sts, seq);
    if (sts == SmppSts.OK) {
      bindResp.systemId(systemId);
      bindResp.addOptionalParam(new Tlv(Tag.SC_INTERFACE_VERSION, new byte[] {0x34}));
    }
    return bindResp;
  }
}

class BindRespPdu extends Pdu {

  private String systemId = "";
  private List<Tlv> optionalParams;

  BindRespPdu(SmppCmd cmd, SmppSts sts, int seq) {
    super(cmd, sts, seq);
  }

  @Override
  byte[] toBytes() {
    byte[] systemIdBytes = systemId.getBytes(StandardCharsets.US_ASCII);
    int len = 16;
    len += (systemIdBytes.length + 1); // plus null terminator
    if (optionalParams != null) {
      for (Tlv p : optionalParams) {
        len += p.lenInBytes();
      }
    }
    var buf = ByteBuffer.allocate(len);
    buf.putInt(len);
    buf.putInt(cmd.getCmdId());
    buf.putInt(sts.getStsId());
    buf.putInt(seq);
    buf.put(systemIdBytes);
    buf.put((byte) 0); // null terminator
    if (optionalParams != null) {
      for (Tlv p : optionalParams) {
        p.writeToBuf(buf);
      }
    }
    return buf.array();
  }

  void addOptionalParam(Tlv tlv) {
    if (optionalParams == null) {
      optionalParams = new ArrayList<>(1);
    }
    optionalParams.add(tlv);
  }

  void systemId(String systemId) {
    if (systemId != null) {
      this.systemId = systemId;
    }
  }
}

class SubmitSm extends Pdu {

  private Addr srcAddr;
  private Addr destAddr;
  private int regDeliv = 0;
  // don't interested in other fields

  SubmitSm(SmppSts sts, int seq) {
    super(SmppCmd.SUBMIT_SM_RESP, sts, seq);
  }

  @Override
  void parseBody(ByteBuffer buf) {
    String serviceType = Utils.parseCStr(buf);
    srcAddr = new Addr(buf.get(), buf.get(), Utils.parseCStr(buf));
    destAddr = new Addr(buf.get(), buf.get(), Utils.parseCStr(buf));
    byte esmClass = buf.get();
    byte protocolId = buf.get();
    byte priority = buf.get();
    String sched = Utils.parseCStr(buf);
    String valid = Utils.parseCStr(buf);
    regDeliv = buf.get();

    // stop parsing (don't interested in other fields)
  }

  Addr srcAddr() {
    return srcAddr;
  }

  Addr destAddr() {
    return destAddr;
  }

  boolean isDlrRequested() {
    return regDeliv != 0;
  }

  SubmitSmResp createResp(SmppSts respSts, String messageId) {
    return new SubmitSmResp(respSts, seq, messageId);
  }
}

class SubmitSmResp extends Pdu {

  private final String messageId;

  SubmitSmResp(SmppSts sts, int seq, String messageId) {
    super(SmppCmd.SUBMIT_SM_RESP, sts, seq);
    this.messageId = messageId;
  }

  @Override
  byte[] toBytes() {
    int len = 16;
    byte[] msgIdBytes = null;
    if (SmppSts.OK == sts && messageId != null) {
      msgIdBytes = messageId.getBytes(StandardCharsets.US_ASCII);
      len += (msgIdBytes.length + 1); // plus 1 for null terminator
    }
    var buf = ByteBuffer.allocate(len);
    buf.putInt(len);
    buf.putInt(cmd.getCmdId());
    buf.putInt(sts.getStsId());
    buf.putInt(seq);
    if (SmppSts.OK == sts && msgIdBytes != null) {
      buf.put(msgIdBytes);
      buf.put((byte) 0); // null terminator
    }
    return buf.array();
  }
}

class DeliverSm extends Pdu {

  private byte[] serviceType = new byte[0];
  private Addr srcAddr;
  private Addr destAddr;
  private byte esmClass;
  private byte dataCoding;
  private byte[] message = new byte[0];
  private List<Tlv> optionalParams = Collections.emptyList();

  DeliverSm(SmppSts sts, int seq) {
    super(SmppCmd.DELIVER_SM, sts, seq);
  }

  void serviceType(String serviceType) {
    this.serviceType = serviceType.getBytes(StandardCharsets.US_ASCII);
  }

  void srcAddr(Addr srcAddr) {
    this.srcAddr = srcAddr;
  }

  void destAddr(Addr destAddr) {
    this.destAddr = destAddr;
  }

  void esmClass(byte esmClass) {
    this.esmClass = esmClass;
  }

  void dataCoding(byte dataCoding) {
    this.dataCoding = dataCoding;
  }

  void message(byte[] message) {
    this.message = message;
  }

  void addOptionalParam(Tlv tlv) {
    if (optionalParams.isEmpty()) {
      optionalParams = new ArrayList<>(1);
    }
    optionalParams.add(tlv);
  }

  @Override
  byte[] toBytes() {
    int len = calcLen();
    var buf = ByteBuffer.allocate(len);
    buf.putInt(len);
    buf.putInt(cmd.getCmdId());
    buf.putInt(sts.getStsId());
    buf.putInt(seq);
    buf.put(serviceType).put((byte) 0); // with null-terminator
    srcAddr.writeToBuf(buf);
    destAddr.writeToBuf(buf);
    buf.put(esmClass);
    buf.put((byte) 0); // protocol id
    buf.put((byte) 0); // priority flag
    buf.put((byte) 0); // schedule delivery time
    buf.put((byte) 0); // validity period
    buf.put((byte) 0); // registered delivery
    buf.put((byte) 0); // replace if present
    buf.put(dataCoding);
    buf.put((byte) 0); // sm_default_msg_id (must be set to null)
    buf.put((byte) message.length);
    buf.put(message);
    for (Tlv tlv : optionalParams) {
      tlv.writeToBuf(buf);
    }
    return buf.array();
  }

  private int calcLen() {
    int len = 16;
    len += (serviceType.length + 1);
    len += srcAddr.lenInBytes();
    len += destAddr.lenInBytes();
    len += 10; // 10 one-byte fields;
    len += message.length;
    for(Tlv tlv : optionalParams) {
      len += tlv.lenInBytes();
    }
    return len;
  }
}

enum SmppCmd {

  GENERIC_NACK(0x80000000),
  BIND_RECEIVER(0x00000001),
  BIND_RECEIVER_RESP(0x80000001),
  BIND_TRANSMITTER(0x00000002),
  BIND_TRANSMITTER_RESP(0x80000002),
  SUBMIT_SM(0x00000004),
  SUBMIT_SM_RESP(0x80000004),
  DELIVER_SM(0x00000005),
  DELIVER_SM_RESP(0x80000005),
  UNBIND(0x00000006),
  UNBIND_RESP(0x80000006),
  BIND_TRANSCEIVER(0x00000009),
  BIND_TRANSCEIVER_RESP(0x80000009),
  ENQUIRE_LINK(0x00000015),
  ENQUIRE_LINK_RESP(0x80000015),
  ;

  private final int cmdId;

  SmppCmd(int cmdId) {
    this.cmdId = cmdId;
  }

  public int getCmdId() {
    return cmdId;
  }

  static SmppCmd fromCmdId(int cmdId) {
    SmppCmd[] vals = SmppCmd.values();
    for (SmppCmd cmd : vals) {
      if (cmd.cmdId == cmdId) {
        return cmd;
      }
    }
    throw new RuntimeException("Unknown smpp command with id " + cmdId);
  }
}

enum SmppSts {
  OK(0x00000000),
  INV_CMD_ID(0x00000003),
  INV_BIND_STS(0x00000004),
  ALREADY_BOUND(0x00000005),
  SYS_ERROR(0x00000008),
  UNKNOWN(0x000000FF),
  ;

  private final int stsId;

  SmppSts(int stsId) {
    this.stsId = stsId;
  }

  int getStsId() {
    return stsId;
  }

  static SmppSts fromStsId(int stsId) {
    for (SmppSts sts : SmppSts.values()) {
      if (sts.stsId == stsId) {
        return sts;
      }
    }
    throw new RuntimeException("Unknown smpp status with id " + stsId);
  }
}

enum Tag {
  PRIVACY_INDICATOR((short) 0x0201),
  MS_VALIDITY((short) 0x1204),
  SC_INTERFACE_VERSION((short) 0x0210),
  RECEIPTED_MSG_ID((short) 0x001e),
  MSG_STATE((short) 0x0427),
  ;

  private final short tagId;

  Tag(short tagId) {
    this.tagId = tagId;
  }

  short getTagId() {
    return tagId;
  }
}

class Tlv {
  private final Tag tag;
  private final short len;
  private final byte[] val;

  Tlv(Tag tag, byte[] val) {
    this.tag = tag;
    this.len = (short) val.length;
    this.val = val;
  }

  int lenInBytes() {
    return 4 + val.length;
  }

  void writeToBuf(ByteBuffer buf) {
    buf.putShort(tag.getTagId());
    buf.putShort(len);
    buf.put(val);
  }
}

record Addr(byte ton, byte npi, String addr) {

  int lenInBytes() {
    int len = 3; // npi + ton + null terminator
    if (addr != null) {
      len += addr.getBytes(StandardCharsets.US_ASCII).length;
    }
    return len;
  }

  void writeToBuf(ByteBuffer buf) {
    buf.put(ton);
    buf.put(npi);
    if (addr != null) {
      buf.put(addr.getBytes(StandardCharsets.US_ASCII));
    }
    buf.put((byte) 0);
  }

  @Override
  public String toString() {
    return ton + ":" + npi + ":" + addr;
  }
}

File: pdu.dart
import 'dart:typed_data';

import 'util.dart';

enum Cmd {
  genericNack(0x80000000),
  bindReceiver(0x00000001),
  bindReceiverResp(0x80000001),
  bindTransmitter(0x00000002),
  bindTransmitterResp(0x80000002),
  bindTransceiver(0x00000009),
  bindTransceiverResp(0x80000009),
  submitSm(0x00000004),
  submitSmResp(0x80000004),
  deliverSm(0x00000005),
  deliverSmResp(0x80000005),
  unbind(0x00000006),
  unbindResp(0x80000006),
  enquireLink(0x00000015),
  enquireLinkResp(0x80000015),
  unknown(0x00000000);

  final int cmdId;

  const Cmd(this.cmdId);

  Cmd toResp() {
    switch(this) {
      case Cmd.bindReceiver:
        return Cmd.bindReceiverResp;
      case Cmd.bindTransceiver:
        return Cmd.bindTransceiverResp;
      case Cmd.bindTransmitter:
        return Cmd.bindTransmitterResp;
      case Cmd.submitSm:
        return Cmd.submitSmResp;
      case Cmd.deliverSm:
        return Cmd.deliverSmResp;
      default:
        throw Exception("Cannot find resp for $this");
    }
  }

  static Cmd fromCmdId(int cmdId) {
    for (var cmd in Cmd.values) {
      if (cmdId == cmd.cmdId) {
        return cmd;
      }
    }
    return Cmd.unknown;
  }
}

enum Sts {
  ok(0x00000000),
  invalidCmd(0x00000003),
  invalidBindSts(0x00000004),
  alreadyBound(0x00000005),
  systemError(0x00000008),
  unknown(-1);

  final int stsId;

  const Sts(this.stsId);

  static Sts fromStsId(int stsId) {
    for (var sts in Sts.values) {
      if (stsId == sts.stsId) {
        return sts;
      }
    }
    return Sts.unknown;
  }
}

enum Tag {
  privacyIndicator(0x0201),
  msValidity(0x1204),
  scInterfaceVersion(0x0210),
  receiptedMessageId(0x001e),
  messageState(0x0427);

  final int _tagCode;

  const Tag(this._tagCode);
}

abstract class Pdu {

  final Cmd cmd;
  final Sts sts;
  final int seq;

  Pdu(this.cmd, this.sts, this.seq);

  void parseBody(ByteData bytes);

  ByteData toBytes();
}

class HeaderPdu extends Pdu {

  HeaderPdu(Cmd cmd, Sts sts, int seq): super(cmd, sts, seq);

  void parseBody(ByteData bytes) {
    // NOOP
  }

  ByteData toBytes() {
    var bytes = ByteData(16);
    bytes.setUint32(0, 16);
    bytes.setUint32(4, cmd.cmdId);
    bytes.setUint32(8, sts.stsId);
    bytes.setUint32(12, seq);
    return bytes;
  }
}

class BindReqPdu extends Pdu {

  String systemId = "";
  // don't interested in other fields

  BindReqPdu(Cmd cmd, Sts sts, int seq): super(cmd, sts, seq);

  @override
  void parseBody(ByteData bytes) {
    String systemId;
    (systemId, _) = bytes.getCString(16);
    this.systemId = systemId;
    // skip parsing other fields
  }

  @override
  ByteData toBytes() {
    throw Exception("unnecessary operation not implemented in smsc");
  }

  BindRespPdu createResp(Sts sts, String? systemId) {
    var resp = BindRespPdu(cmd.toResp(), sts, seq);
    resp.systemId = systemId;
    return resp;
  }
}

class BindRespPdu extends Pdu {

  String? systemId;

  BindRespPdu(Cmd cmd, Sts sts, int seq): super(cmd, sts, seq);

  @override
  void parseBody(ByteData bytes) {
    throw Exception("unnecessary operation not implemented in smsc");
  }

  @override
  ByteData toBytes() {
    var interfaceVersion = Tlv.byte(Tag.scInterfaceVersion, 0x34);
    var len = 16 + interfaceVersion._lenInBytes();
    if (Sts.ok == sts) {
      // also need space for systemId
      len += (systemId != null ? (systemId!.length + 1) : 0);
    }
    int offset = 0;
    var bytes = ByteData(len);
    bytes.setUint32(offset + 0, len);
    bytes.setUint32(offset + 4, cmd.cmdId);
    bytes.setUint32(offset + 8, sts.stsId);
    bytes.setUint32(offset + 12, seq);
    offset = 16;
    if (Sts.ok == sts && systemId != null) {
      offset = bytes.setCString(offset, systemId);
    }
    interfaceVersion._writeToBytes(bytes, offset);
    return bytes;
  }
}

class Addr {
  final String addr;
  final int ton;
  final int npi;

  Addr(this.addr, this.ton, this.npi);

  int _lenInBytes() {
    return (addr.length + 1) + 2; // +1 for null terminator
  }

  int _writeToBytes(ByteData b, int offset) {
    b.setUint8(offset++, ton);
    b.setUint8(offset++, npi);
    return b.setCString(offset, addr);
  }
}

class Tlv {
  final Tag tag;
  final int len;
  final ByteData? val;

  Tlv(Tag tag, ByteData? val):
        this.tag = tag,
        this.len = val?.lengthInBytes ?? 0,
        this.val = val;

  Tlv.byte(Tag tag, int val):
      this.tag = tag,
      this.len = 1,
      this.val = _byteVal(val);

  int _lenInBytes() {
    var len = 4;
    len += val?.lengthInBytes ?? 0;
    return len;
  }

  int _writeToBytes(ByteData b, int offset) {
    b.setUint16(offset, tag._tagCode);
    offset += 2;
    b.setUint16(offset, len);
    offset += 2;
    return b.setBytes(offset, val);
  }

  static ByteData _byteVal(int n) {
    var b = new ByteData(1);
    b.setUint8(0, n);
    return b;
  }
}

class SubmitSmPdu extends Pdu {

  Addr? srcAddr;
  Addr? destAddr;
  int regDeliv = 0;
  // don't interested in other fields

  SubmitSmPdu(Sts sts, int seq): super(Cmd.submitSm, sts, seq);

  @override
  void parseBody(ByteData bytes) {
    int offset = 16;
    (_, offset) = bytes.getCString(offset); // service_type
    int srcTon = bytes.getUint8(offset);
    int srcNpi = bytes.getUint8(offset+1);
    String src;
    (src, offset) = bytes.getCString(offset + 2);
    int destTon = bytes.getUint8(offset);
    int destNpi = bytes.getUint8(offset+1);
    String dst;
    (dst, offset) = bytes.getCString(offset + 2);
    (_, offset) = bytes.getCString(offset + 3); // sched_time (skipped esm_class, protocol_id & priority class)
    (_, offset) = bytes.getCString(offset); // validity_period
    this.regDeliv = bytes.getUint8(offset);
    this.srcAddr = Addr(src, srcTon, srcNpi);
    this.destAddr = Addr(dst, destTon, destNpi);
  }

  @override
  ByteData toBytes() {
    throw Exception("unnecessary operation not implemented in smsc");
  }

  SubmitSmRespPdu createResp(Sts sts, String? messageId) {
    var resp = SubmitSmRespPdu(sts, seq);
    resp.messageId = messageId;
    return resp;
  }
}

class SubmitSmRespPdu extends Pdu {

  String? messageId;

  SubmitSmRespPdu(Sts sts, int seq): super(Cmd.submitSmResp, sts, seq);

  @override
  void parseBody(ByteData bytes) {
    throw Exception("unnecessary operation not implemented in smsc");
  }

  @override
  ByteData toBytes() {
    var len = 16;
    if (Sts.ok == sts) {
      // also need space for message_id
      len = len + (messageId != null ? (messageId!.length + 1) : 0);
    }
    var bytes = ByteData(len);
    bytes.setUint32(0, len);
    bytes.setUint32(4, cmd.cmdId);
    bytes.setUint32(8, sts.stsId);
    bytes.setUint32(12, seq);
    if (Sts.ok == sts && messageId != null) {
      int offset = 16;
      bytes.setCString(offset, messageId);
    }
    return bytes;
  }
}

class DeliverSmPdu extends Pdu {

  String serviceType = "";
  Addr? srcAddr;
  Addr? destAddr;
  int esmClass = 0;
  int dataCoding = 0;
  ByteData? message;
  List<Tlv> optionalParams = List.empty(growable: true);

  DeliverSmPdu(Sts sts, int seq): super(Cmd.deliverSm, sts, seq);

  @override
  void parseBody(ByteData bytes) {
    throw Exception("unnecessary operation not implemented in smsc");
  }

  @override
  ByteData toBytes() {
    var len = _calcLen();
    var b = ByteData(len);
    b.setUint32(0, len);
    b.setUint32(4, cmd.cmdId);
    b.setUint32(8, sts.stsId);
    b.setUint32(12, seq);
    int offset = 16;
    offset = b.setCString(offset, serviceType);
    if (srcAddr != null) {
      offset = srcAddr!._writeToBytes(b, offset);
    } else {
      offset += 3;
    }
    if (destAddr != null) {
      offset = destAddr!._writeToBytes(b, offset);
    } else {
      offset += 3;
    }
    b.setUint8(offset++, esmClass);
    b.setUint8(offset++, 0); // protocol id
    b.setUint8(offset++, 0); // priority flag
    b.setUint8(offset++, 0); // schedule delivery time
    b.setUint8(offset++, 0); // validity period
    b.setUint8(offset++, 0); // registered delivery
    b.setUint8(offset++, 0); // replace if present
    b.setUint8(offset++, dataCoding);
    b.setUint8(offset++, 0); // sm_default_msg_id (must be set to null)
    b.setUint8(offset++, message?.lengthInBytes ?? 0);
    offset = b.setBytes(offset, message);
    for (var tlv in optionalParams) {
      offset = tlv._writeToBytes(b, offset);
    }
    return b;
  }

  int _calcLen() {
    var len = 16;
    len += (serviceType.length + 1); // c-string
    len += srcAddr?._lenInBytes() ?? 3;
    len += destAddr?._lenInBytes() ?? 3;
    len += 10; // ten 1-byte smpp fields
    len += message?.lengthInBytes ?? 0;
    for (var tlv in optionalParams) {
      len += tlv._lenInBytes();
    }
    return len;
  }
}
